subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/MultiReleaseJarTransformForLegacy.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.classpath.transforms;
+
+import org.gradle.internal.classpath.ClasspathBuilder;
+import org.gradle.internal.classpath.ClasspathEntryVisitor;
+import org.gradle.internal.classpath.ClasspathWalker;
+import org.gradle.internal.classpath.types.InstrumentingTypeRegistry;
+
+import java.io.File;
+import java.io.IOException;
+
+import static org.gradle.internal.classpath.transforms.MrJarUtils.isInUnsupportedMrJarVersionedDirectory;
+
+/**
+ * Transformation for legacy instrumentation when transformed JARs are part of the classpath.
+ * This is still used when TestKit and TAPI run Gradle in embedded or debug mode.
+ * <p>
+ * This transformation filters out not yet supported versioned directories of the multi-release JARs.
+ */
+public class MultiReleaseJarTransformForLegacy extends BaseJarTransform {
```

Comment:
:nail_care: can `MultiReleaseJarTransformForLegacy` be a package-private class?

Reply:
It can be, fixed